From: Pascal Nowack <Pascal.Nowack@gmx.de>
Date: Fri, 20 Mar 2026 06:34:02 +0100
Subject: rdp-pipewire-stream: Only try to use sync timelines when using
 dma-bufs

gnome-remote-desktop currently enforces sync timeline objects even when
using MemFds. However, they can only be used with dma-bufs, which is
currently not the case when using the NVIDIA driver, resulting in a
non-functioning screencast. Fix this issue by checking whether dma-bufs
can actually be used and only if successful negotiate sync timelines.

Origin: https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/merge_requests/394
---
 src/grd-rdp-pipewire-stream.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/grd-rdp-pipewire-stream.c b/src/grd-rdp-pipewire-stream.c
index 5e33b9c..eae08d0 100644
--- a/src/grd-rdp-pipewire-stream.c
+++ b/src/grd-rdp-pipewire-stream.c
@@ -496,6 +496,7 @@ on_stream_param_changed (void                 *user_data,
   g_autoptr (GArray) pod_offsets = NULL;
   struct spa_pod_dynamic_builder pod_builder;
   enum spa_data_type allowed_buffer_types;
+  gboolean use_explicit_sync = FALSE;
   struct spa_pod_frame buffers_frame;
   g_autoptr (GPtrArray) params = NULL;
 
@@ -536,6 +537,10 @@ on_stream_param_changed (void                 *user_data,
   if (egl_thread && !hwaccel_nvidia)
     allowed_buffer_types |= 1 << SPA_DATA_DmaBuf;
 
+  if (allowed_buffer_types & 1 << SPA_DATA_DmaBuf &&
+      grd_egl_thread_supports_explicit_sync (egl_thread))
+    use_explicit_sync = TRUE;
+
   grd_append_pod_offset (pod_offsets, &pod_builder.b);
   spa_pod_builder_push_object (&pod_builder.b, &buffers_frame,
                                SPA_TYPE_OBJECT_ParamBuffers, SPA_PARAM_Buffers);
@@ -544,7 +549,7 @@ on_stream_param_changed (void                 *user_data,
     SPA_PARAM_BUFFERS_buffers, SPA_POD_CHOICE_RANGE_Int (8, 2, 8),
     SPA_PARAM_BUFFERS_dataType, SPA_POD_Int (allowed_buffer_types),
     0);
-  if (egl_thread && grd_egl_thread_supports_explicit_sync (egl_thread))
+  if (use_explicit_sync)
     {
       spa_pod_builder_prop (&pod_builder.b,
                             SPA_PARAM_BUFFERS_metaType,
@@ -569,7 +574,7 @@ on_stream_param_changed (void                 *user_data,
                                                    CURSOR_META_SIZE (1, 1),
                                                    CURSOR_META_SIZE (384, 384)));
 
-  if (egl_thread && grd_egl_thread_supports_explicit_sync (egl_thread))
+  if (use_explicit_sync)
     {
       grd_append_pod_offset (pod_offsets, &pod_builder.b);
       spa_pod_builder_add_object
